### PR TITLE
layers: add ICMPv6 NDP

### PR DESCRIPTION
**Pull request by jcrussell**
Implementing decoders for the neighbor discovery protocol. Decoding seems to be working on a couple of simple test cases:

```
=== RUN   TestPacketICMPv6RouterAdvertisement
--- PASS: TestPacketICMPv6RouterAdvertisement (0.00s)
        base_test.go:25: Checking packet layers, want [Ethernet IPv6 ICMPv6 ICMPv6RouterAdvertisement]
        base_test.go:27:   Got layer Ethernet, 14 bytes, payload of 104 bytes
        base_test.go:27:   Got layer IPv6, 40 bytes, payload of 64 bytes
        base_test.go:27:   Got layer ICMPv6, 4 bytes, payload of 60 bytes
        base_test.go:27:   Got layer ICMPv6RouterAdvertisement, 0 bytes, payload of 0 bytes
        base_test.go:30: PACKET: 118 bytes
                - Layer 1 (14 bytes) = Ethernet {Contents=[..14..] Payload=[..104..] SrcMAC=c2:00:54:f5:00:00 DstMAC=33:33:00:00:00:01 EthernetType=IPv6 Length=0}
                - Layer 2 (40 bytes) = IPv6     {Contents=[..40..] Payload=[..64..] Version=6 TrafficClass=224 FlowLabel=0 Length=64 NextHeader=ICMPv6 HopLimit=255 SrcIP=fe80::c000:54ff:fef5:0 DstIP=ff02::1 HopByHop=nil}
                - Layer 3 (04 bytes) = ICMPv6   {Contents=[134, 0, 196, 254] Payload=[..60..] TypeCode=RouterAdvertisement Checksum=50430}
                - Layer 4 (00 bytes) = ICMPv6RouterAdvertisement        {Contents=[] Payload=[] HopLimit=64 Flags=0 RouterLifetime=1800 ReachableTime=0 RetransTimer=0 Options=[{Type=1 Length=8 Data=[..6..]}, {Type=5 Length=8 Data=[..6..]}, {Type=3 Length=32 Data=[..30..]}]}

=== RUN   TestPacketICMPv6NeighborSolicitation
--- PASS: TestPacketICMPv6NeighborSolicitation (0.00s)
        base_test.go:25: Checking packet layers, want [Ethernet IPv6 ICMPv6 ICMPv6NeighborSolicitation]
        base_test.go:27:   Got layer Ethernet, 14 bytes, payload of 64 bytes
        base_test.go:27:   Got layer IPv6, 40 bytes, payload of 24 bytes
        base_test.go:27:   Got layer ICMPv6, 4 bytes, payload of 20 bytes
        base_test.go:27:   Got layer ICMPv6NeighborSolicitation, 0 bytes, payload of 0 bytes
        base_test.go:30: PACKET: 78 bytes
                - Layer 1 (14 bytes) = Ethernet {Contents=[..14..] Payload=[..64..] SrcMAC=00:0c:29:0e:4c:67 DstMAC=33:33:ff:0e:4c:67 EthernetType=IPv6 Length=0}
                - Layer 2 (40 bytes) = IPv6     {Contents=[..40..] Payload=[..24..] Version=6 TrafficClass=0 FlowLabel=0 Length=24 NextHeader=ICMPv6 HopLimit=255 SrcIP=:: DstIP=ff02::1:ff0e:4c67 HopByHop=nil}
                - Layer 3 (04 bytes) = ICMPv6   {Contents=[135, 0, 185, 48] Payload=[..20..] TypeCode=NeighborSolicitation Checksum=47408}
                - Layer 4 (00 bytes) = ICMPv6NeighborSolicitation       {Contents=[] Payload=[] TargetAddress=fe80::20c:29ff:fe0e:4c67 Options=[]}
```

Comments welcome, I'm not sure that this is the right approach.

### CODE REVIEW

File: layers/icmp6.go [hunk 1]
```diff
@@ -165,9 +165,8 @@ func CreateICMPv6TypeCode(typ uint8, code uint8) ICMPv6TypeCode {
 // ICMPv6 is the layer for IPv6 ICMP packet data
 type ICMPv6 struct {
 	BaseLayer
-	TypeCode  ICMPv6TypeCode
-	Checksum  uint16
-	TypeBytes []byte
```
Comment (gconnell) [inline layers/icmp6.go]: Please don't remove TypeBytes, as that's a backwards-incompatible change which could break code of anyone currently referencing the field.

**Comment by gconnell** [comment]
Thanks for the code!  Please make sure the .travis.golint.sh script at the top-level directory succeeds, as well (probably it's complaining about comments on structs).

File: layers/icmp6.go (hunk 1 above)
Comment (jcrussell) [reply layers/icmp6.go]: Alright, I'll add it back to the struct. Should I keep the old behavior of setting it to the `data[4:8]` or leave it empty?

**Comment by jcrussell** [comment]
Can I change `checkLayers` so that if you have `Payload` in want, it accepts any layer type in that index and ignore layers beyond that? For example:

```
--- FAIL: TestICMP6 (0.00s)
base_test.go:25: Checking packet layers, want [Ethernet IPv6 ICMPv6 Payload]
base_test.go:27:   Got layer Ethernet, 14 bytes, payload of 72 bytes
base_test.go:27:   Got layer IPv6, 40 bytes, payload of 32 bytes
base_test.go:27:   Got layer ICMPv6, 4 bytes, payload of 28 bytes
base_test.go:27:   Got layer ICMPv6NeighborSolicitation, 0 bytes, payload of 0 bytes
...
base_test.go:38:   Layer 3 mismatch: got ICMPv6NeighborSolicitation want Payload
```

**Comment by jcrussell** [comment]
@gconnell: do you think it's worth adding structs to decode all the values for the various options? For example:

```
4.6.2.  Prefix Information

       0                   1                   2                   3
       0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
      |     Type      |    Length     | Prefix Length |L|A| Reserved1 |
      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
      |                         Valid Lifetime                        |
      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
      |                       Preferred Lifetime                      |
      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
      |                           Reserved2                           |
      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
      |                                                               |
      +                                                               +
      |                                                               |
      +                            Prefix                             +
      |                                                               |
      +                                                               +
      |                                                               |
      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

With what I've written so far, users would be left with a `[]byte`.

If so, how would you implement it? One possibility would be to add a `AsPrefix` function to the `ICMPv6Option` struct that returns this (or an error if the option is the wrong type)? It could be nice to have a struct for each type of option that implements `DecodeFromBytes` and `SerializeTo`.

**Comment by gconnell** [comment]
I'd definitely be cool with a struct... you can look at a couple of the
other protos that have optional stuff (even TCP options, IPv6 optional
stuff, DNS records, etc) and see what they do.

**Comment by jcrussell** [comment]
I decided to follow the example in `dhcpv4.go` and `tcp.go` -- add Stringers instead structs for each option.

**Comment by jcrussell** [comment]
@gconnell: any chance you can take another look at this PR?

**Pull request merged**